link-verifier/verify-links.py

```diff
@@ -0,0 +1,397 @@
+        epilog='Requires beautifulsoup4, requests, and termcolor from PyPi. ' +
+               'Optional dependencies: pandoc (to support testing Markdown files), gh (To speed up checking GitHub links)'
+    )
+    parser.add_argument("-F", "--files", action="store", dest="files", nargs='+', help="List of files to test links in.")
+    parser.add_argument("-L", "--links", action="store", dest="links", nargs='+', help="List of links to test.")
+    parser.add_argument("-D", "--exclude-dirs", action="store", dest="exclude_dirs", nargs='+', help="List of directories to ignore.")
+    parser.add_argument("-I", "--include-file-types", action="store", dest="include_files", nargs='+', help="List of file patterns to search for URLs.")
+    parser.add_argument("-A", "--allowlist-file", action="store", dest="allowlist", help="Path to file containing list of allowed URLs.")
+    parser.add_argument("-n", "--num-processes", action="store", type=int, default=4, help="Number of processes to run in parallel")
+    parser.add_argument("-k", "--keep", action="store_true", default=False, help="Keep temporary files instead of deleting")
+    parser.add_argument("-v", "--verbose", action="store_true", default=False, help="Print all links tested")
+    args = parser.parse_args()
+
+    html_file_list = []
+    broken_links = []
+    file_list = []
+    link_list = []
+    exclude_dirs = [dir.lower() for dir in args.exclude_dirs] if args.exclude_dirs else []
+
+    print(exclude_dirs)
+
+    # If any explicit files are passed, add them to file_list.
+    if args.files is not None:
+        file_list = args.files
+    else:
+        # Obtain list of Markdown files from the repository (along with excluding passed directories).
+        for root, dirs, files in os.walk("./"):
+            # Prune dirs to remove exclude directories, if passed as command line input, from search.
+            dirs[:] = [dir for dir in dirs if dir.lower() not in exclude_dirs]
+            file_list += [os.path.join(root, f) for f in files if re.search(MARKDOWN_SEARCH_TERM, f, re.IGNORECASE)]
+
+    if args.verbose:
+        print(file_list)
+
+    # If any explicit links are passed, add them to link_list.
+    if args.links is not None:
+        link_list = args.links
+    elif args.include_files is not None:
+        for root, dirs, files in os.walk("./"):
+            # Avoid exclude directories, if passed, from search.
+            dirs[:] = [dir for dir in dirs if dir.lower() not in exclude_dirs]
+            for file in files:
+                if any(file.endswith(file_type) for file_type in args.include_files):
+                    text = open(os.path.join(root, file), 'r').read()
+                    print(os.path.join(root, file))
+                    urls = re.findall(URL_SEARCH_TERM, text)
+                    for url in urls:
+                        if url[0] not in link_list:
+                            link_list.append(url[0])
```

Comment:
Any file that we open should also be closed. I'd prefer using a `with` as that will close on exceptions too. Also, if printing the file is only useful for debugging, we probably shouldn't do it every time and put it behind `args.verbose`
```suggestion
                    with open(os.path.join(root, file), 'r') as f:
                        text = f.read()
                        if args.verbose:
                            print(f.name)
                        urls = re.findall(URL_SEARCH_TERM, text)
                        for url in urls:
                            if url[0] not in link_list:
                                link_list.append(url[0])
```

Reply:
I will update the code to use `with`. The printing of files was only for local testing, we don't necessarily need to print it even with the `--verbose` option.